From fb475a3c2a4e5ed7271cf4952e7b84c85ab6e20d Mon Sep 17 00:00:00 2001
From: Bo-Cun Chen <bc-bocun.chen@mediatek.com>
Date: Wed, 24 Sep 2025 12:15:16 +0800
Subject: [PATCH] netfilter: add DEV_PATH_MTK_WDMA path to xt_FLOWOFFLOAD for
 the RelayD

Without this patch, nf_dev_path_info will populate an incorrect source
MAC address while running RelayD. Consequently, traffic may experience
stalls in both the software fastpath and the hardware path.

Signed-off-by: Bo-Cun Chen <bc-bocun.chen@mediatek.com>
---
 net/netfilter/xt_FLOWOFFLOAD.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/net/netfilter/xt_FLOWOFFLOAD.c b/net/netfilter/xt_FLOWOFFLOAD.c
index ab56262..98c3872 100644
--- a/net/netfilter/xt_FLOWOFFLOAD.c
+++ b/net/netfilter/xt_FLOWOFFLOAD.c
@@ -396,6 +396,10 @@ static void nf_dev_path_info(const struct net_device_path_stack *stack,
 				break;
 			}
 			break;
+		case DEV_PATH_MTK_WDMA:
+			if (is_zero_ether_addr(info->h_source))
+				memcpy(info->h_source, path->dev->dev_addr, ETH_ALEN);
+			break;
 		default:
 			break;
 		}
-- 
2.45.2

